File: Algorithms/p089_Gray_Code/p089_Gray_Code.py
class Solution(object):
    def grayCode(self, n):
        """
        :type n: int
        :rtype: List[int]
        """
        upper_bound = 2**n
        res = [0]*upper_bound
        for i in range(0,upper_bound):
            for j in range(0,n):
                res[i] |= ((i>>1)&(0x1<<j))^(i&(0x1<<j))
        return res

File: Algorithms/p089_Gray_Code/p089_Gray_Code.cpp
class Solution {
public:
    vector<int> grayCode(int n) {
        int upper_bound = 1;
        for(int i=0;i<n;++i){
            upper_bound *= 2;
        }
        vector<int> res(upper_bound);
        for(int i=0;i<upper_bound;++i){
            for(int j=0;j<n;++j){
                res[i] |= ((i>>1)&(0x1<<j))^(i&(0x1<<j));
            }
        }
        return res;
    }
};
